karton/gateway/operations.py

```diff
@@ -0,0 +1,464 @@
+        TaskState.STARTED,
+        TaskState.FINISHED,
+        TaskState.CRASHED,
+    ):
+        return True
+    if old_status is TaskState.STARTED and new_status in (
+        TaskState.FINISHED,
+        TaskState.CRASHED,
+    ):
+        return True
+    return False
+
+
+@request_handler
+async def handle_set_task_status_request(
+    websocket: WebSocket, request: SetTaskStatusRequest, session: UserSession
+) -> None:
+    task_token = request.message.token
+    task_info = parse_task_token(
+        task_token, gateway_config.secret_key, session.user.username
+    )
+    service_backend = await session.get_service_backend()
+    task = await service_backend.get_task(task_info.task_uid)
+    if task is None:
+        raise InvalidTaskError("Task no longer exists")
+
+    new_task_status = request.message.status
+    if not is_valid_task_status_transition(
+        old_status=task.status, new_status=new_task_status
+    ):
+        raise InvalidTaskStatusError(
+            f"Can't change task status from {task.status.value}"
+            f" to {new_task_status.value}"
+        )
+
+    if new_task_status is TaskState.CRASHED:
+        task.error = request.message.error
+        await service_backend.increment_metrics(
+            KartonMetrics.TASK_CRASHED, session.identity
+        )
+
+    await service_backend.set_task_status(task, new_task_status)
+    await service_backend.increment_metrics(
+        KartonMetrics.TASK_CONSUMED, session.identity
+    )
+    await send_success(websocket)
+
+
+async def generate_resource_download_urls(
+    task: Task, service_backend: KartonAsyncBackend
+) -> list[ResourceUrl]:
+    resources = {}
+
+    for resource in task.iterate_resources():
+        if resource.uid in resources:
+            continue
+        if resource.bucket != service_backend.default_bucket_name:
+            raise InvalidTaskError(
+                "Got task that references different bucket than the configured one "
+                "which is unsupported"
+            )
+        download_url = await service_backend.get_presigned_object_download_url(
+            bucket=service_backend.default_bucket_name, object_uid=resource.uid
+        )
+        resources[resource.uid] = ResourceUrl(uid=resource.uid, url=download_url)
+    return list(resources.values())
+
+
+@request_handler
+async def handle_get_task_request(
+    websocket: WebSocket, request: GetTaskRequest, session: UserSession
+) -> None:
+    if not session.is_bound:
+        raise InvalidBindError("Service has not declared consumer bind")
+
+    identity = session.identity
+    service_backend = await session.get_service_backend()
+
+    current_bind = service_backend.get_bind(identity)
+    if current_bind != session.karton_bind:
+        raise ExpiredBindError(
+            f"Consumer bind expired. "
+            f"Old binds: {str(session.karton_bind)}. "
+            f"New binds: {str(current_bind)}."
```

Comment:
I think this is equivalent to
```suggestion
            f"Old binds: {session.karton_bind}. "
            f"New binds: {current_bind}."
```